level05/CoffeeMachineTests.hs

```diff
@@ -0,0 +1,336 @@
+newtype SetDrinkType (v :: Type -> Type) = SetDrinkType DrinkType deriving Show
+
+instance HTraversable AddMilkSugar where
+  htraverse _ (AddMilkSugar d) = pure $ AddMilkSugar d
+
+instance HTraversable SetDrinkType where
+  htraverse _ (SetDrinkType d) = pure $ SetDrinkType d
+
+cSetDrinkType
+  :: forall g m. (MonadGen g, MonadTest m, MonadIO m)
+  => C.Machine
+  -> Command g m Model
+cSetDrinkType mach = Command gen exec
+  [ Update $ \m (SetDrinkType d) _ -> m
+    & modelDrinkType .~ d
+    & modelMilk .~ 0
+    & modelSugar .~ 0
+    & modelSufficientCredit .~ TooLittle
+
+  , Ensure $ \_ m _ drink -> case (m ^. modelDrinkType, drink) of
+      (Coffee, C.Coffee{})           -> success
+      (HotChocolate, C.HotChocolate) -> success
+      (Tea, C.Tea{})                 -> success
+      _                              -> failure
+  ]
+  where
+    gen :: Model Symbolic -> Maybe (g (SetDrinkType Symbolic))
+    gen _ = Just $ SetDrinkType <$> Gen.enumBounded
+
+    exec :: SetDrinkType Concrete -> m C.Drink
+    exec (SetDrinkType d) = evalIO $ do
+      mach & case d of
+        Coffee       -> C.coffee
+        HotChocolate -> C.hotChocolate
+        Tea          -> C.tea
+      view C.drinkSetting <$> C.peek mach
+
+milkOrSugar :: DrinkAdditive -> a -> a ->  a
+milkOrSugar Milk m  _ = m
+milkOrSugar Sugar _ s = s
+
+milkOrSugarExec
+  :: ( MonadTest m
+     , MonadIO m
+     )
+  => C.Machine
+  -> AddMilkSugar Concrete
+  -> m C.Drink
+milkOrSugarExec mach (AddMilkSugar additive) = do
+  evalIO $ milkOrSugar additive C.addMilk C.addSugar mach
+  view C.drinkSetting <$> C.peek mach
+
+genAddMilkSugarCommand
+  :: MonadGen g
+  => (DrinkType -> Bool)
+  -> Model Symbolic
+  -> Maybe (g (AddMilkSugar Symbolic))
+genAddMilkSugarCommand isDrinkType m
+  | isDrinkType (m ^. modelDrinkType) = Just (AddMilkSugar <$> Gen.enumBounded)
+  | otherwise                         = Nothing
+
+cAddMilkSugarSad
+  :: forall g m. (MonadGen g, MonadTest m, MonadIO m)
+  => C.Machine
+  -> Command g m Model
+cAddMilkSugarSad mach = Command (genAddMilkSugarCommand (== HotChocolate)) (milkOrSugarExec mach)
+  [ Require $ \m _ ->
+      doesntHaveMug m || m ^. modelDrinkType == HotChocolate
+
+  , Ensure $ \_ _ _ drink ->
+      drink ^? C._HotChocolate === Just ()
+  ]
+
+cAddMilkSugarHappy
+  :: forall g m. (MonadGen g, MonadTest m, MonadIO m)
+  => C.Machine
+  -> Command g m Model
+cAddMilkSugarHappy ref = Command (genAddMilkSugarCommand (/= HotChocolate)) (milkOrSugarExec ref)
+  [ Require $ \m _ ->
+      hasMug m && m ^. modelDrinkType /= HotChocolate
+
+  , Update $ \m (AddMilkSugar additive) _ -> m
+      & milkOrSugar additive modelMilk modelSugar +~ 1
+      & modelSufficientCredit .~ TooLittle
+
+  , Ensure $ \oldM newM (AddMilkSugar additive) mug' ->
+      let (mL, sl) = milkOrSugar additive (modelMilk, C.milk) (modelSugar, C.sugar)
+
+          drinkAdditiveQty = case newM ^. modelDrinkType of
+            Coffee -> mug' ^? C._Coffee . sl
```

Comment:
I used `Control.Lens.failing` to glue the two prisms together. you may want to do so